src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/GoogleDataprocDAO.scala

```diff
@@ -1,79 +1,244 @@
     new GoogleCredential.Builder()
       .setTransport(httpTransport)
       .setJsonFactory(jsonFactory)
       .setServiceAccountId(dataprocConfig.serviceAccount)
-      .setServiceAccountScopes(cloudPlatformScopes.asJava)
-      .setServiceAccountPrivateKeyFromPemFile(new java.io.File(dataprocConfig.serviceAccountPemPath))
+      .setServiceAccountScopes(scopes.asJava)
+      .setServiceAccountPrivateKeyFromPemFile(serviceAccountPemFile)
       .build()
   }
 
+  /* Creates a cluster in the given google project:
+       - Add a firewall rule to the user's google project if it doesn't exist, so we can access the cluster
+       - Create the initialization bucket for the cluster in the leo google project
+       - Upload all the necessary initialization files to the bucket
+       - Create the cluster in the google project
+     Currently, the bucketPath of the clusterRequest is not used - it will be used later as a place to store notebook results */
+  def createCluster(googleProject: GoogleProject, clusterName: String, clusterRequest: ClusterRequest)(implicit executionContext: ExecutionContext): Future[ClusterResponse] = {
+    // Create the firewall rule in the google project if it doesn't already exist, so we can access the cluster
+    updateFirewallRules(googleProject)
 
-  def createCluster(googleProject: String, clusterName: String, clusterRequest: ClusterRequest)(implicit executionContext: ExecutionContext): Future[ClusterResponse] = {
-    val op = build(googleProject, clusterName, clusterRequest)
-    op.map{op =>
-      val metadata = op.getMetadata
-      ClusterResponse(clusterName, googleProject, metadata.get("clusterUuid").toString, metadata.get("status").toString, metadata.get("description").toString, op.getName)}
+    val bucketName = s"${clusterName}-${UUID.randomUUID.toString}"
+    // Create the bucket in leo's google bucket and populate with initialization files
+    val bucketResponse = initializeBucket(dataprocConfig.leoGoogleBucket, clusterName, bucketName)
+    // Once the bucket is ready, build the cluster
+    bucketResponse.flatMap[ClusterResponse] { _ =>
+      val operation = build(googleProject, clusterName, clusterRequest, bucketName)
+      // Once the cluster creation request is sent to Google, it returns a DataprocOperation, which we transform into a ClusterResponse
+      operation.map { op =>
+        val metadata = op.getMetadata
+        ClusterResponse(clusterName, googleProject, metadata.get("clusterUuid").toString, metadata.get("status").toString, metadata.get("description").toString, op.getName)
+      }
+    }
+  }
+
+  /* Adds a firewall rule in the given google project if it does not already exist. */
+  private def updateFirewallRules(googleProject: String) = {
+    val request = new Compute(httpTransport, jsonFactory, getServiceAccountCredential(vmScopes)).firewalls().get(googleProject, dataprocConfig.clusterFirewallRuleName)
```

Comment:
Should we instantiate a `Compute.Builder` as a lazy val, like we do with `Storage` and `Dataproc`?